riscv: use riscv,isa-extensions dt prop

maass-hamburg opened the pull request:
implement and use riscv,isa-extensions
dt prop, like in linux
https://www.kernel.org/doc/Documentation/devicetree/bindings/riscv/extensions.yaml
to set the riscv extensions.

Usefull for softcores like vexriscv or vexiiriscv, that can be easily configured to add these extentions and we don't want to create a separate soc for every possible combination.

For now just just applying it to the litex_vexriscv soc, applying it for the whole tree will be part of a separate PR.

maass-hamburg committed:
riscv: use riscv,isa-extensions dt prop

implement and use riscv,isa-extensions
dt prop, like in linux
https://www.kernel.org/doc/Documentation/devicetree/bindings/riscv/extensions.yaml
to set the riscv extentions.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
riscv: make riscv,isa dt prop no longer required

make riscv,isa dt prop no longer required,
as it is is not really used by anyrhing in zephyr
and we now have a alternative (riscv,isa-base
and riscv,isa-extensions).

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
litex: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for
litex vexriscv soc.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

fkokosinski reviewed (commented):
This is good stuff, but I'd love to see this applied to the rest of RISC-V SoCs in Zephyr.

maass-hamburg commented:
> This is good stuff, but I'd love to see this applied to the rest of RISC-V SoCs in Zephyr.

@fkokosinski 
Do we want to apply it now in this PR and probably risk this not being merged before the feature freeze this Friday,
or have this in version 4.3 and migrate the whole tree, as same as making this mandatory, for 4.4?

mbolivar commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -888,17 +888,51 @@ def dt_nodelabel_array_prop_has_val(kconf, _, label, prop, val):
+    except edtlib.EDTError:
+        return "n"
+
+    if not node or (prop not in node.props):
```
The function bodies introduced in this patch have a fair bit of duplicated code. Please factor out the common code into a single place.

mbolivar reviewed (changes requested):
No objections to the idea, some comments about the specifics. @fkokosinski I'm not sure about some of your comments; I've responded to them individually -- PTAL!

maass-hamburg commented on dts/bindings/cpu/riscv,cpus.yaml:
```diff
@@ -17,3 +17,11 @@ properties:
+
+  riscv,isa-base:
+    description: The base ISA implemented by the hart.
+    type: string
```
in zephyr we also have `rv32e` used and a Kconfig option for `rv128i`, but that's not used in tree

mbolivar replied:
Can you explain the divergence from Linux here?

maass-hamburg commented on dts/bindings/cpu/riscv,cpus.yaml:
```diff
@@ -15,5 +15,12 @@ properties:
-    required: true
     type: string
+
+  riscv,isa-base:
```
If we migrate from selecting the riscv extension and base from the soc, to only using the dt as the source, they have to be set, as `riscv,isa` is not used by zephyr, while Linux has support for parsing both.

maass-hamburg commented:
rebased due to conflicts

fkokosinski commented on dts/bindings/cpu/riscv,cpus.yaml:
```diff
@@ -15,5 +15,12 @@ properties:
 
   riscv,isa:
```
Hmm, the main difference is that Linux is still using the property (at least in checks and informing the user that the property is indeed deprecated, e.g. here: https://github.com/torvalds/linux/blob/dcb6fa37fd7bc9c3d2b066329b0d27dedf8becaa/arch/riscv/kernel/cpu.c#L100), whereas Zephyr doesn't use this property at all.

Not sure if we want to keep a completely unused property in the tree?

maass-hamburg committed:
qemu: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for
qemu riscv boards.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

mbolivar reviewed (commented):
@fkokosinski I've resolved almost all my comments -- I think the remaining question is whether to

1. remove riscv,isa and mark riscv,isa-base and riscv,isa-extensions as required or
2. deprecate riscv,isa and leave the other two as not required

Did I get that right?

fkokosinski commented:
@mbolivar - 

> Did I get that right?

Correct.

You've made a good point [here](https://github.com/zephyrproject-rtos/zephyr/pull/97540#discussion_r2466520464) about OOT dts, which convinces me we should go with deprecation first in 4.4, and then with removal in e.g. 4.5 to give OOT some time to adjust.

rruuaanng commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -940,25 +940,64 @@ def dt_nodelabel_enabled_with_compat(kconf, _, label, compat):
+        return "n"
+
+    if prop not in node.props:
+        return "n"
```
```suggestion
    if node is None or prop not in node.props:
        return "n"
```

maass-hamburg committed:
riscv: deprecate riscv,isa dt prop

deprecate riscv,isa dt prop

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
efinix: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for
efinix riscv boards.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
renode: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for
renode riscv boards.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg commented:
`/cpus/cpu@0` is not used for all riscv cpus (f.e. nordic doesn't, see zephyr/dts/vendor/nordic/nrf54lm20a.dtsi). This is bad because we already expect it at other places like `config DCACHE_LINE_SIZE` for `dt_node_int_prop_int,/cpus/cpu@0,d-cache-line-size`.

Should we
- require that the cpu we are building for are always at `/cpus/cpu@0`
- use a nodelabel instead (and require that)
- use the compat `riscv`

?

maass-hamburg committed:
scripts: Add list array support for dt_compat_any_has_prop

Add support for arrays and string-arrays to
dt_compat_any_has_prop
kconfig preprocessor function.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
scripts: Add dt_compat_all_has_prop

Add the dt_compat_all_has_prop
kconfig preprocessor function, simillar to
dt_compat_any_has_prop.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
scripts: Refactor dt_compat_any_has_prop

Refactor dt_compat_any_has_prop
to reduce its Cognitive Complexity.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
riscv: dts: add missing "riscv" compatible

Add the "riscv" compatible, where it was
missing.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
arch: riscv: require "riscv" compatible

Require the "riscv" compatible for
CONFIG_RISCV.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
docs: migration-guide: riscv compatible requirement

mention riscv compatible requirement.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
nordic: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for
nordic riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
espressif: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for
espressif riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
wch: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for
wch riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
telink: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for
telink riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
starfive: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for
starfive riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
snps: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for
snps riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
sifive: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for
sifive riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
bflb: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for
bflb riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
sensry: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for
sensry riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
raspberrypi: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for
raspberrypi riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
openisa: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
openhwgroup: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for riscv cpus.

also combines both SOC_CV64A6 variants

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
neorv32: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
microchip: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
lowriscv: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
ite: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
intel: niosv: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
gd: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
egis: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
andestech: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
aesc: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
thead: riscv: use riscv,isa-extensions dt prop

use riscv,isa-extensions dt prop for riscv cpus.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
docs: migration-guide: mention riscv changes

mention riscv changes.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
starfive: riscv: dts: remove unsupported properties

Remove unsupported properties from the StarFive RISC-V DTS files.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
riscv: require `riscv,isa-base` and `riscv,isa-extensions`

require `riscv,isa-base` and `riscv,isa-extensions` dt props
to be set.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

fkokosinski reviewed (approved):
Thanks! LGTM
